godot-core/src/task/async_runtime.rs

```diff
@@ -0,0 +1,456 @@
+        self.value = FutureSlotState::Gone;
+    }
+
+    /// Attempts to extract the future with the given ID from the slot.
+    ///
+    /// Puts the slot into [`FutureSlotState::Polling`] state after taking the future out. It is expected that the future is either parked
+    /// again or the slot is cleared.
+    /// In cases were the slot state is not [`FutureSlotState::Pending`], a copy of the state is returned but the slot remains untouched.
+    fn take_for_polling(&mut self, id: u64) -> FutureSlotState<T> {
+        match self.value {
+            FutureSlotState::Empty => FutureSlotState::Empty,
+            FutureSlotState::Polling => FutureSlotState::Polling,
+            FutureSlotState::Gone => FutureSlotState::Gone,
+            FutureSlotState::Pending(_) if self.id != id => FutureSlotState::Gone,
+            FutureSlotState::Pending(_) => {
+                std::mem::replace(&mut self.value, FutureSlotState::Polling)
+            }
+        }
+    }
+
+    /// Parks the future in this slot again.
+    ///
+    /// # Panics
+    /// - If the slot is not in state [`FutureSlotState::Polling`].
+    fn park(&mut self, value: T) {
+        match self.value {
+            FutureSlotState::Empty | FutureSlotState::Gone => {
+                panic!("Future slot is currently unoccupied, future can not be parked here!");
+            }
+            FutureSlotState::Pending(_) => {
+                panic!("Future slot is already occupied by a different future!")
+            }
+            FutureSlotState::Polling => {
+                self.value = FutureSlotState::Pending(value);
+            }
+        }
+    }
+}
+
+#[derive(Default)]
+/// The storage for the pending tasks of the async runtime.
+struct AsyncRuntime {
+    tasks: Vec<FutureSlot<Pin<Box<dyn Future<Output = ()>>>>>,
+    next_task_id: u64,
+    #[cfg(feature = "trace")]
+    panicked_tasks: std::collections::HashSet<u64>,
+}
+
+impl AsyncRuntime {
+    fn new() -> Self {
+        Self {
+            tasks: Vec::with_capacity(10),
+            next_task_id: 0,
+            #[cfg(feature = "trace")]
+            panicked_tasks: std::collections::HashSet::default(),
+        }
+    }
+
+    /// Get the next task ID.
+    fn next_id(&mut self) -> u64 {
+        let id = self.next_task_id;
+        self.next_task_id += 1;
+        id
+    }
+
+    /// Store a new async task in the runtime.
+    ///
+    /// First, a linear search is performed to locate an already existing but currently unoccupied slot in the task buffer. If there is no
+    /// free slot, a new slot is added which may grow the underlying [`Vec`].
+    ///
+    /// The future storage always starts out with a capacity of 10 tasks.
+    fn add_task<F: Future<Output = ()> + 'static>(&mut self, future: F) -> TaskHandle {
+        let id = self.next_id();
+        let index_slot = self
+            .tasks
+            .iter_mut()
+            .enumerate()
+            .find(|(_, slot)| slot.is_empty());
```

Comment:
`iter()` instead of `iter_mut()` would be enough here, no?

Reply:
Yes 👍 good catch.

Reply:
We actually mutate the found slot in line 293. I added a comment about that to make it more obvious.